File: app/src/main/java/com/heavymagikhq/lbc_player_ultimate/lbcActivity/view/LBCPlayerActivity.java
package com.heavymagikhq.lbc_player_ultimate.lbcActivity.view;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.google.android.exoplayer2.ui.SimpleExoPlayerView;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.LbcPlayerActivityPresenter;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.model.beans.LbcShow;
import com.heavymagikhq.lbc_player_ultimate.R;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.showsRV.ShowSelectedOnClickListener;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.showsRV.ShowsRVAdapter;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.daysRV.DayRVAdapter;
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.daysRV.DaySelectedOnClickListener;
import com.heavymagikhq.lbc_player_ultimate.lbcExoPlayer.LbcExoPlayer;
import com.heavymagikhq.lbc_player_ultimate.lbcStreamingService.RadioStreamingService;
import com.heavymagikhq.lbc_player_ultimate.lbcStreamingService.RadioStreamingServiceBinder;

import org.jetbrains.annotations.NotNull;

import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class LBCPlayerActivity extends AppCompatActivity implements DaySelectedOnClickListener, ShowSelectedOnClickListener, LbcPlayerView {

    public static final String PODCAST_URI_KEY = "KEY";
    public static final String liveLbcUrl = "http://media-ice.musicradio.com/LBCLondonMP3";
    private final String TAG = LBCPlayerActivity.class.getSimpleName();
    private LbcPlayerActivityPresenter lbcPlayerActivityPresenter;

    @BindView(R.id.liveLbcPlayer)
    SimpleExoPlayerView liveLbcExoPlayerView;
    @BindView(R.id.podtcastLbcShowPlayer)
    SimpleExoPlayerView podcastExoPlayerView;
    @BindView(R.id.daySelectionRV)
    RecyclerView dateSelectionRV;
    @BindView(R.id.showSelectionRV)
    RecyclerView showSelectionRV;
    @BindView(R.id.swipeRL)
    SwipeRefreshLayout swipeRL;

    private DayRVAdapter datesRVAdapter = null;
    private ShowsRVAdapter showsRVAdapter = null;
    private RadioStreamingService radioStreamingService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        startExoPlayerService();
        bindService(new Intent(this, RadioStreamingService.class), connectionToRadioStreamingService, Context.BIND_AUTO_CREATE);
        initRVAdapters();
        initPresenter();
        initSwipeRL();
    }

    //MARK: ---------- INITIALISATION

    private void initPresenter() {
        lbcPlayerActivityPresenter = new LbcPlayerActivityPresenter(this);
    }

    private void initSwipeRL() {
        final int green = getResources().getColor(android.R.color.holo_green_dark);
        final int blue = getResources().getColor(R.color.colorPrimaryDark);
        swipeRL.setColorSchemeColors(green, blue);
        swipeRL.setOnRefreshListener(this::onRefreshShowsData);
    }

    private void initRVAdapters() {
        datesRVAdapter = new DayRVAdapter();
        showsRVAdapter = new ShowsRVAdapter();
        dateSelectionRV.setAdapter(datesRVAdapter);
        showSelectionRV.setAdapter(showsRVAdapter);
    }

    private void startExoPlayerService() {
        final Intent startExoPlayerServiceIntent = new Intent(this, RadioStreamingService.class);
        startExoPlayerServiceIntent.putExtra(PODCAST_URI_KEY, LBCPlayerActivity.liveLbcUrl);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            startForegroundService(startExoPlayerServiceIntent);
        else
            startService(new Intent(startExoPlayerServiceIntent));
    }

    /**
     * binds the the liveLbcExoPlayerView to the radioStreamingService
     *
     * @param radioStreamingService - the service which contains the exoPlayer which plays LBC
     */
    private void bindExoPlayerToExoPlayerView(LbcExoPlayer radioStreamingService) {
        liveLbcExoPlayerView.setPlayer(radioStreamingService.getExoPlayer());
    }

    private ServiceConnection connectionToRadioStreamingService = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            if (service instanceof RadioStreamingServiceBinder) {
                radioStreamingService = ((RadioStreamingServiceBinder) service).getRadioStreamingService();
                bindExoPlayerToExoPlayerView((radioStreamingService.getLiveLbcExoPlayer()));
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
        }
    };

    @Override
    public void onSetDatesTo(@NotNull List<? extends Date> dates, int selectedDateIndex) {
        datesRVAdapter.setDates(dates);
        datesRVAdapter.setSelectedDateIndex(selectedDateIndex);
        dateSelectionRV.getLayoutManager().scrollToPosition(selectedDateIndex);
        showsRVAdapter.notifyDataSetChanged();
    }

    //MARK: ---------- INITIALISATION

    //MARK: ---------- SHOW SELECTION

    @Override
    public void onDateSelected(Date selectedDate, int selectedPosition) {
        lbcPlayerActivityPresenter.onDateSelected(selectedDate);
        datesRVAdapter.setSelectedDateIndex(selectedPosition);
        datesRVAdapter.notifyDataSetChanged();
    }

    @Override
    public void onShowSelected(@NotNull LbcShow show, int selectedPosition) {
        if (podcastExoPlayerView.getPlayer() != null)
            podcastExoPlayerView.getPlayer().release();

        final LbcExoPlayer podcastExoPlayer = new LbcExoPlayer(this, show.getStreamUrl());
        podcastExoPlayerView.setPlayer(podcastExoPlayer.getExoPlayer());
        radioStreamingService.setPodcastPlayer(podcastExoPlayer);

        showsRVAdapter.setSelectedShow(show);
        showsRVAdapter.notifyDataSetChanged();
    }

    @Override
    public void listThese(@NotNull List<LbcShow> shows) {
        runOnUiThread(() -> {
            showsRVAdapter.setLbcShows(shows);
            showsRVAdapter.notifyDataSetChanged();

            showSelectionRV
                    .getLayoutManager()
                    .scrollToPosition(
                            showsRVAdapter.getIndexOfShowToSelected()
                    );

            if (swipeRL.isRefreshing())
                swipeRL.setRefreshing(false);
        });
    }


    @Override
    public void openDownloadBrowserFor(@NotNull LbcShow show, int selectedPosition) {
        final Intent launchPodcastWebsiteIntent = new Intent(Intent.ACTION_VIEW);
        final Uri showUri = Uri.parse(show.getStreamUrl());
        launchPodcastWebsiteIntent.setData(showUri);
        startActivity(launchPodcastWebsiteIntent);
    }

    //MARK: ---------- SHOW SELECTION

    //MARK: ---------- SHOW REFRESHING & ERROR HANDLING

    private void onRefreshShowsData() {
        final Date selectedDate = datesRVAdapter.getSelectedDate();
        final Date currentDate = new Date();
        if (selectedDate != null)
            lbcPlayerActivityPresenter.retrieveShowsThenSelect(selectedDate);
        else
            lbcPlayerActivityPresenter.retrieveShowsThenSelect(currentDate);
    }

    @Override
    public void onHandleFailureToGetShows(@NotNull String errorMessage) {
        runOnUiThread(() -> {
                    Toast.makeText(this, errorMessage, Toast.LENGTH_LONG).show();
                    if (swipeRL.isRefreshing())
                        swipeRL.setRefreshing(false);
                }
        );
    }

    //MARK: ---------- SHOW REFRESHING & ERROR HANDLING

    //MARK: ---------- LIFECYCLE

    @Override
    protected void onDestroy() {
        super.onDestroy();
        //free resources:
        unbindService(connectionToRadioStreamingService);
        stopService(new Intent(this, RadioStreamingService.class));
        if (podcastExoPlayerView.getPlayer() != null)
            podcastExoPlayerView.getPlayer().release();
    }

    //MARK: ---------- LIFECYCLE
}

File: app/src/main/java/com/heavymagikhq/lbc_player_ultimate/lbcActivity/view/LBCPlayerActivity.kt
package com.heavymagikhq.lbc_player_ultimate.lbcActivity.view


import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.View.OnFocusChangeListener
import android.widget.EditText
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import com.google.android.gms.cast.framework.CastButtonFactory
import com.heavymagikhq.lbc_player_ultimate.R
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.LbcPlayerActivityPresenter
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.model.services.lbcShowRetrieverService.LbcPresenter
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.model.services.lbcShowRetrieverService.LbcShow
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.headersRV.PresenterSelectedOnClickListener
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.headersRV.PresentersRVAdapter
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.showsRV.ShowSelectedOnClickListener
import com.heavymagikhq.lbc_player_ultimate.lbcActivity.view.showsRV.ShowsRVAdapter
import com.heavymagikhq.lbc_player_ultimate.lbcExoPlayer.LbcExoPlayer
import kotlinx.android.synthetic.main.activity_main.*


class LBCPlayerActivity : AppCompatActivity(), PresenterSelectedOnClickListener,
    ShowSelectedOnClickListener, LbcPlayerView {
    private lateinit var showsRVAdapter: ShowsRVAdapter
    private lateinit var presenterRvAdapter: PresentersRVAdapter
    private lateinit var liveExoPlayer: LbcExoPlayer
    private var podcastExoPlayer: LbcExoPlayer? = null
    private lateinit var lbcPlayerActivityPresenter: LbcPlayerActivityPresenter

    companion object {
        const val liveLbcUrl = "http://media-ice.musicradio.com/LBCLondonMP3"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        liveExoPlayer = LbcExoPlayer(liveLbcUrl, this, true, liveLbcPlayer, podcastExoPlayer)

        initRVAdapters()
        initPresenter()
        initSwipeRL()
        initPresenterSearchBar()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu, menu)
        CastButtonFactory.setUpMediaRouteButton(this, menu, R.id.media_route_menu_item)
        return true
    }

    // MARK: ---------- INITIALISATION

    @RequiresApi(Build.VERSION_CODES.R)
    private fun initPresenterSearchBar() {
        presenterSearchBar.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                return true
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                presenterRvAdapter.presenterFilter.filter(newText)
                return true
            }
        })

        presenterSearchBar.findViewById<EditText>(R.id.search_src_text).onFocusChangeListener =
            OnFocusChangeListener { v, hasFocus ->
                if (!hasFocus) {
                    presenterSearchBar.isIconified = true
                    presenterRV.layoutManager?.scrollToPosition(
                        presenterRvAdapter.selectedPresenterIdx
                    )
                }
            }
    }

    private fun initPresenter() {
        lbcPlayerActivityPresenter = LbcPlayerActivityPresenter(this)
    }

    private fun initSwipeRL() {
        val green = resources.getColor(android.R.color.holo_green_dark)
        val blue = resources.getColor(R.color.colorPrimaryDark)
        swipeRL!!.setColorSchemeColors(green, blue)
        swipeRL!!.setOnRefreshListener { onRefreshShowsData() }
    }

    private fun initRVAdapters() {
        showsRVAdapter = ShowsRVAdapter()
        showSelectionRV!!.adapter = showsRVAdapter
    }

    // MARK: ---------- INITIALISATION

    // MARK: ---------- PRESENTER SELECTION

    override fun onPresentersSet(presenters: List<LbcPresenter>) {
        runOnUiThread {
            presenterRvAdapter = PresentersRVAdapter(presenters, presenterRV!!.layoutManager!!)
            presenterRV!!.adapter = presenterRvAdapter
            presenterRvAdapter.notifyDataSetChanged()
        }
    }

    override fun onPresenterSelected(presenter: LbcPresenter, selectedPosition: Int) {
        val prevSelectedIdx = presenterRvAdapter.selectedPresenterIdx
        presenterRvAdapter.selectedPresenter = presenter
        presenterRvAdapter.notifyItemChanged(prevSelectedIdx)
        presenterRvAdapter.notifyItemChanged(selectedPosition)
        lbcPlayerActivityPresenter.loadShows(presenter, false)
    }

    override fun listPresenterShows(shows: List<LbcShow>) {
        runOnUiThread {
            showsRVAdapter.lbcShows = shows
            showsRVAdapter.notifyDataSetChanged()
            showSelectionRV!!
                .layoutManager!!
                .scrollToPosition(
                    showsRVAdapter.indexOfSelectedShow
                )
            if (swipeRL!!.isRefreshing) swipeRL!!.isRefreshing = false
        }
    }

    // MARK: ---------- PRESENTER SELECTION

    // MARK: ---------- SHOW SELECTION

    override fun openDownloadBrowserFor(show: LbcShow, selectedPosition: Int) {
        val launchPodcastWebsiteIntent = Intent(Intent.ACTION_VIEW)
        val showUri = Uri.parse(show.streamUrl)
        launchPodcastWebsiteIntent.data = showUri
        startActivity(launchPodcastWebsiteIntent)
    }

    override fun onShowSelected(show: LbcShow, selectedPosition: Int) {
        liveExoPlayer.pauseLBC()

        val playShow = {
            podcastExoPlayer =
                LbcExoPlayer(show.playableUrl, this, false, podcastPlayer, liveExoPlayer)
            liveExoPlayer.otherPlayer = podcastExoPlayer
            showsRVAdapter.selectedShow = show
            showsRVAdapter.notifyDataSetChanged()
        }

        if (showsRVAdapter.selectedShow != null && showsRVAdapter.selectedShow != show) {
            val builder = AlertDialog.Builder(this)
            builder.apply {
                setPositiveButton(R.string.ok) { _, _ -> playShow() }
                setNegativeButton(R.string.cancel, null)
            }
            builder.create().show()
        } else {
            playShow()
        }
    }

    // MARK: ---------- SHOW SELECTION

    // MARK: ---------- SHOW REFRESHING & ERROR HANDLING

    private fun onRefreshShowsData() {
        if (showsRVAdapter.presenter != null)
            lbcPlayerActivityPresenter.loadShows(showsRVAdapter.presenter!!, true)
        else
            swipeRL.isRefreshing = false
    }

    override fun onHandleFailureToGetShows(errorMessage: String) {
        runOnUiThread {
            Toast.makeText(this, errorMessage, Toast.LENGTH_LONG).show()
            if (swipeRL.isRefreshing) swipeRL.isRefreshing = false
        }
    }

    // MARK: ---------- SHOW REFRESHING & ERROR HANDLING

    // MARK: ---------- LIFECYCLE

    override fun onDestroy() {
        super.onDestroy()
        liveExoPlayer.release()
        podcastExoPlayer?.release()
    }

    // MARK: ---------- LIFECYCLE
}